source/developer_manual/design/processes.rst

```diff
@@ -0,0 +1,146 @@
+can be prefixed with the process's name and the username of the user who
+was logged in, without the possibility for the process itself to forge
+them. Flooding logs can also be prevented. By default Dovecot allows
+non-privileged processes to write 10 lines per second before it begins
+to delay reading their input, which finally causes the badly behaving
+process to start blocking on writing to stderr instead of eating all the
+CPU and disk space.
+
+In the Dovecot 2.0 design, the master process is split to three parts:
+the Master process which does nothing more than keep the processes
+running, the config process which handles reading the configuration file
+(supporting also eg. SQL storages!) and the log process which handles
+the logging.
+
+
+Login processes
+---------------
+
+The login processes implement the required minimum of the IMAP and POP3
+protocols before a user logs in successfully. There are separate
+processes (and binaries) to handle IMAP and POP3 protocols.
+
+These processes are run with least possible privileges. Unfortunately
+the default UNIX security model still allows them to do much more than
+they would have to: Accept new connections on a socket, connect to new
+UNIX sockets and read and write to existing file descriptors. Still, the
+login process is by default run under a user account that has no special
+access to anything, and inside a non-writable chroot where only a couple
+of files exist. Doing any damage inside there should be difficult.
+
+When a new connection comes, one of the login processes accept()s it.
+After that the client typically does nothing more than ask the server's
+capability list and then log in. The client may also start TLS session
+before logging in.
+
+Authentication is done by talking to the authentication process. The
+login process is completely untrusted by the authentication process, so
+even if an attacker is able to execute arbitrary code inside a login
+process, they won't be able to log in without a valid username and
+password.
+
+After receiving a successful authentication reply from the
+authentication process, the login process sends the file descriptor to
+the master process which creates a new mail process and transfers the fd
+into it. Before doing that, the master process verifies from the
+authentication process that the authentication really was successful.
+
+By default each login process will handle only a single connection and
+afterwards kill itself (but see SSL proxying below). This way attacker
+can't see other people's connections. This can however be disabled
+(``login_process_per_connection=no``), in which case the security of the
+design suffers greatly.
+
+The login processes handle SSL/TLS connections themselves completely.
+They keep proxying the connection to mail processes for the entire
+lifetime of the connection. This way if a security hole is found from
+the SSL library, an authenticated user still can't execute code outside
+the login process.
```

Comment:
It can create outgoing internet connections though and thereby compromise the network from within if this server has direct access to DMZ.

This remains to be fixed.